File: challenge_22/challenge_22.exs
defmodule Score do
  def names(file) do
    case File.read(file) do
      {:ok, body}     ->
        all(body)
      {:error,reason} ->
        :file.format_error(reason)
    end
  end

  def all(body) do
    body
      |> String.replace(~r/\W+/, ",")
      |> String.split(",", trim: true)
      |> Stream.map(&(String.to_char_list(&1)))
      |> Enum.sort()
      |> count(0,1)
  end

  def count(name,total,_) when length(name) == 0, do: total
  def count([name|names],total,i) when length(names) >= 0 do
    count(names,(total + char_prod(name,i)),i+1)
  end

  def char_prod(char_list,m), do: char_sums(char_list) * m
  def char_sums(char_list) do
    char_list
      |> Stream.map(fn y -> y - 64 end)
      |> Enum.reduce(fn b,c -> b + c end)
  end
end

IO.inspect Score.names("challenge_22/challenge_22_names.txt")

File: challenge_22/challenge_22.rb
name_file = File.open("challenge_22/challenge_22_names.txt",'r').read.gsub(/\W+/, ',').split(',').sort
alphabet_hash = Hash[("A".."Z").to_a.zip((1..26).to_a)]
letters = name_file.map { |name| name.split("").map { |letter| alphabet_hash[letter] } }.map { |n| n.inject(:+) }
letters.each_with_index { |name,index| letters[index] = name.to_i*(index) }
p letters.inject(:+)
